### PR TITLE
Speed up FSRS inference, evaluation, and training without production Burn dependency

### PR DESCRIPTION
**Pull request by asukaminato0721**
**Summary**
This branch removes Burn from the production FSRS path and replaces the tensor/autodiff-based implementation with scalar and analytic implementations. Burn remains available as a dev-dependency for oracle tests, but normal library builds no longer depend on it. The branch also improves evaluation, memory-state calculation, simulation workload estimation, and adds benchmarks/tests for the new `card_ids` training path.

**Key Improvements**
- Removed Burn from normal dependencies, reducing release library size and build footprint.
- Reimplemented core FSRS inference with scalar state transitions for `memory_state`, `memory_state_batch`, retrievability, intervals, and next-state calculations.
- Added analytic loss and gradient computation for parameter training, replacing production autodiff.
- Added a host-side Adam optimizer and training loop with progress/abort handling.
- Added optional `card_ids` to `ComputeParametersInput` so training can group same-card prefixes into one full card trajectory and avoid repeated recurrence work.
- Optimized `memory_state_batch`, including a fast path for repeated identical inputs.
- Improved simulation workload estimation by replacing ndarray matrices with flat storage, interpolation helpers, cached transitions, and in-flight card cost caching.
- Fixed disabled short-term training initialization so short-term parameters are zeroed/frozen consistently.
- Added verification and perf tooling for Burn removal and autoresearch-style correctness/benchmark checks.

**Performance Highlights**
From the included perf comparison artifact:

- `parameters/evaluate`: about `5.8x` faster.
- `parameters/evaluate_with_time_series_splits`: about `10.2x` faster.
- `benchmark/next_states`: about `281x` faster.
- `benchmark/calc_mem`: about `222x` faster.
- `benchmark/calc_mem_batch`: roughly `35x` to `59x` faster depending on case.
- `parameters/compute_parameters`: about `3.0x` faster without `card_ids`, and about `9.3x` faster with `card_ids`.

Build artifact impact from local release-lib measurement:

- `libfsrs-*.rlib`: reduced from about `3.49 MB` to about `1.14 MB`, around `67%` smaller.
- `target/release/deps`: reduced by about `286 MB`, around `76%`.
- Clean release-lib build time dropped from about `32s` to about `8s`.

**Testing**
Validated with:

```sh
SKIP_TRAINING=1 cargo test
cargo bench --bench parameters -- compute_parameters
```

Additional coverage includes analytic-vs-Burn oracle tests, host Adam tests, disabled short-term training tests, windowed card-prefix loss/gradient equivalence tests, and `compute_parameters` comparisons with and without `card_ids`.

**Acknowledgements**
Many of the optimization directions in this PR were inspired by [Expertium/fsrs-rs-speed-autoresearch](https://github.com/Expertium/fsrs-rs-speed-autoresearch), which explored performance improvements for `fsrs-rs` through autoresearch-style experimentation. Thanks to @Expertium for making those findings available and helping inform this implementation.

### CODE REVIEW

**Comment by Expertium** [comment]
Hey, I'm glad that you found my autoresearch repo useful! 👍
That being said, I think more speedups from my repo should be ported. `compute_parameters` was sped up by 100x (200x with AVX2), that's way more than 3-9x. If you want to be conservative and don't want to change too much all at once - that's fine, just keep in mind that there are more speedups available.
EDIT: [Claude just updated the readme in my repo to clarify how much improvement came from exact vs inexact changes](https://github.com/Expertium/fsrs-rs-speed-autoresearch/#how-much-of-the-104-was-free-vs-a-precision-trade). Personally, I recommend keeping everything other than O(N^2) -> O(N) FSRSItems thing. That one increased log loss significantly. Without that change a ~30-35x speedup should be available.

File: src/simulation.rs [hunk 1]
```diff
@@ -535,41 +698,41 @@ impl WorkloadEstimator {
...
                         current_cost += (review_costs[rating - 1] + future_cost) * transition_prob;
                     }
-                    unsafe {
-                        *cost_matrix.uget_mut([s_idx, d_idx, t]) = current_cost;
-                    }
+                    let cost_index = self.layer_offset(t) + self.state_offset(s_idx, d_idx);
+                    cost_matrix[cost_index] = current_cost;
                 }
             }
         }
```
[the remark this answers is not in the document]
Comment (L-M-Sherlock) [reply src/simulation.rs:751]: it has been done on L739.

File: src/training.rs [hunk 2]
```diff
@@ -1,68 +1,48 @@
...
+    retrievability: &[f32],
+    labels: &[f32],
+    weights: &[f32],
+) -> f32 {
+    let mut loss = 0.0;
+    let mut weight_sum = 0.0;
+    for ((&r, &label), &weight) in retrievability.iter().zip(labels).zip(weights) {
+        loss += (label * r.ln() + (1.0 - label) * (1.0 - r).ln()) * weight;
+        weight_sum += weight;
     }
```
[the remark this answers is not in the document]
Comment (L-M-Sherlock) [reply src/training.rs:39]: But this might mask other issues during the training process.

**Comment by asukaminato0721** [comment]
cannot replicate the speedup of SIMD

**Review by L-M-Sherlock — APPROVED** [review]
Great Work!
